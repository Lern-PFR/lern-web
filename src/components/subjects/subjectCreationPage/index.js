import SubjectCreationForm from './SubjectCreationForm';

export {
	// eslint-disable-next-line import/prefer-default-export
	SubjectCreationForm,
};
