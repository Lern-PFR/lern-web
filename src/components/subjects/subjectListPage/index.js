import SubjectCard from './SubjectCard';

export {
	// eslint-disable-next-line import/prefer-default-export
	SubjectCard,
};
