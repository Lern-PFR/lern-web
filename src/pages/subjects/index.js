import SubjectCreationPage from './SubjectCreationPage';
import SubjectDetailsPage from './SubjectDetailsPage';
import SubjectListPage from './SubjectListPage';

export {
	SubjectCreationPage,
	SubjectDetailsPage,
	SubjectListPage,
};
