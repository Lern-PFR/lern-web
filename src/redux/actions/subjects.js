import * as SubjectsApi from 'api/subjectsApi';

/**
 * @constant
 * @name ActionTypes
 * @description The various action types used to interact with the subjects redux state.
 * @type {object}
 *
 * @author Timothée Simon-Franza
 */
export const ActionTypes = {
	FETCH_SUBJECT_REQUEST: '@SUBJECTS/FETCH_REQUEST',
	FETCH_SUBJECT_SUCCESS: '@SUBJECTS/FETCH_SUCCESS',
	FETCH_SUBJECT_FAILURE: '@SUBJECTS/FETCH_FAILURE',

	FETCH_SUBJECT_LIST_REQUEST: '@SUBJECTS/FETCH_LIST_REQUEST',
	FETCH_SUBJECT_LIST_SUCCESS: '@SUBJECTS/FETCH_LIST_SUCCESS',
	FETCH_SUBJECT_LIST_FAILURE: '@SUBJECTS/FETCH_LIST_FAILURE',

	CREATE_SUBJECT_REQUEST: '@SUBJECTS/CREATE_REQUEST',
	CREATE_SUBJECT_SUCCESS: '@SUBJECTS/CREATE_SUCCESS',
	CREATE_SUBJECT_FAILURE: '@SUBJECTS/CREATE_FAILURE',

	UPDATE_SUBJECT_REQUEST: '@SUBJECTS/UPDATE_REQUEST',
	UPDATE_SUBJECT_SUCCESS: '@SUBJECTS/UPDATE_SUCCESS',
	UPDATE_SUBJECT_FAILURE: '@SUBJECTS/UPDATE_FAILURE',

	DELETE_SUBJECT_REQUEST: '@SUBJECTS/DELETE_REQUEST',
	DELETE_SUBJECT_SUCCESS: '@SUBJECTS/DELETE_SUCCESS',
	DELETE_SUBJECT_FAILURE: '@SUBJECTS/DELETE_FAILURE',

	CLEAR_SUBJECT_LIST: '@SUBJECTS/CLEAR_LIST',
};

// //////////////////////////////////////////////////////// //
// /////////// Specific subject fetching actions ////////// //
// //////////////////////////////////////////////////////// //

/**
 * @function
 * @name fetchSubjectRequest
 * @description Action triggered anytime a subject fetching call is made to the API.
 *
 * @author Timothée Simon-Franza
 *
 * @returns {object}
 */
const fetchSubjectRequest = () => ({ type: ActionTypes.FETCH_SUBJECT_REQUEST });

/**
 * @function
 * @name fetchSubjectSuccess
 * @description Action triggered as a result to a successful subject fetching API call.
 *
 * @author Timothée Simon-Franza
 *
 * @param {object} subject : The subject retrieved from the API.
 *
 * @returns {object}
 */
const fetchSubjectSuccess = ({ subject }) => ({
	type: ActionTypes.FETCH_SUBJECT_SUCCESS,
	payload: { subject },
});

/**
 * @function
 * @name fetchSubjectFailure
 * @description Action triggered as a result to a failed subject fetching API call.
 *
 * @author Timothée Simon-Franza
 *
 * @param {object} error : The exception sent back from the API.
 *
 * @returns {object}
 */
const fetchSubjectFailure = (error) => ({
	type: ActionTypes.FETCH_SUBJECT_FAILURE,
	payload: { error },
});

// //////////////////////////////////////////////////////// //
// ///////////// Subject list fetching actions //////////// //
// //////////////////////////////////////////////////////// //

/**
 * @function
 * @name fetchSubjectListRequest
 * @description Action triggered anytime a subject list fetching call is made to the API.
 *
 * @author Timothée Simon-Franza
 *
 * @returns {object}
 */
const fetchSubjectListRequest = () => ({ type: ActionTypes.FETCH_SUBJECT_LIST_REQUEST });

/**
 * @function
 * @name fetchSubjectListSuccess
 * @description Action triggered as a result to a successful subject list fetching API call.
 *
 * @author Timothée SimonFranza
 *
 * @param {array}	subjects	: The list of subjects retrieved from the API.
 * @param {number}	totalCount	: The total amount of subjects available in the database to the current user.
 *
 * @returns {object}
 */
const fetchSubjectListSuccess = ({ subjects, totalCount }) => ({
	type: ActionTypes.FETCH_SUBJECT_LIST_SUCCESS,
	payload: { subjects, totalCount },
});

/**
 * @function
 * @name fetchSubjectListFailure
 * @description Action triggered as a result to a failed subject list fetching API call.
 *
 * @author Timothée Simon-Franza
 *
 * @param {objcet} error : The exception sent back from the API.
 *
 * @returns {object}
 */
const fetchSubjectListFailure = (error) => ({
	type: ActionTypes.FETCH_SUBJECT_LIST_FAILURE,
	payload: { error },
});

// //////////////////////////////////////////////////////// //
// /////////////// Subject creation actions /////////////// //
// //////////////////////////////////////////////////////// //

/**
 * @function
 * @name createSubjectRequest
 * @description Action triggered anytime a subject creation call is made to the API.
 *
 * @author Timothée Simon-Franza
 *
 * @returns {object}
 */
const createSubjectRequest = () => ({ type: ActionTypes.CREATE_SUBJECT_REQUEST });

/**
 * @function
 * @name createSubjectSuccess
 * @description Action triggered as a result to a successfult subject creation API call.
 *
 * @author Timothée Simon-Franza
 *
 * @param {object} subject : The newly created subject.
 *
 * @returns {object}
 */
const createSubjectSuccess = ({ subject }) => ({
	type: ActionTypes.CREATE_SUBJECT_SUCCESS,
	payload: { subject },
});

/**
 * @function
 * @name createSubjectFailure
 * @description Action triggered as a result to a failed subject creation API call.
 *
 * @author Timothée Simon-Franza
 *
 * @param {object} error : The exception sent back from the API.
 *
 * @returns {object}
 */
const createSubjectFailure = (error) => ({
	type: ActionTypes.CREATE_SUBJECT_FAILURE,
	payload: { error },
});

// //////////////////////////////////////////////////////// //
// //////////////// Subject edition actions /////////////// //
// //////////////////////////////////////////////////////// //

/**
 * @function
 * @name updateSubjectRequest
 * @description Action triggered anytime a subject edition call is made to the API.
 *
 * @author Timothée Simon-Franza
 *
 * @returns {object}
 */
const updateSubjectRequest = () => ({ type: ActionTypes.UPDATE_SUBJECT_REQUEST });

/**
 * @function
 * @name updateSubjectSuccess
 * @description Action triggered as a result to a successful subject edition API call.
 *
 * @author Timothée Simon-Franza
 *
 * @param {object} subject : The updated subject object.
 *
 * @returns {object}
 */
const updateSubjectSuccess = ({ subject }) => ({
	type: ActionTypes.UPDATE_SUBJECT_SUCCESS,
	payload: { subject },
});

/**
 * @function
 * @name updateSubjectFailure
 * @description Action triggered as a result to a failed subject edition API call.
 *
 * @author Timothée Simon-Franza
 *
 * @param {object} error : The exception sent back from the API.
 *
 * @returns {object}
 */
const updateSubjectFailure = (error) => ({
	type: ActionTypes.UPDATE_SUBJECT_FAILURE,
	payload: { error },
});

// //////////////////////////////////////////////////////// //
// /////////////// Subject deletion actions /////////////// //
// //////////////////////////////////////////////////////// //

/**
 * @function
 * @name deleteSubjectRequest
 * @description Action triggered anytime a subject deletion call is made to the API.
 *
 * @author Timothée Simon-Franza
 *
 * @returns {object}
 */
const deleteSubjectRequest = () => ({ type: ActionTypes.DELETE_SUBJECT_REQUEST });

/**
 * @function
 * @name deleteSubjectSuccess
 * @description Action triggered as a result to a successful subject deletion API call.
 *
 * @author Timothée Simon-Franza
 *
 * @param {object} subject : The deleted subject object.
 *
 * @returns {object}
 */
const deleteSubjectSuccess = ({ subject }) => ({
	type: ActionTypes.DELETE_SUBJECT_SUCCESS,
	payload: { subject },
});

/**
 * @function
 * @name deleteSubjectFailure
 * @description Action triggered as a result to a failed subject deletion API call.
 *
 * @author Timothée Simon-Franza
 *
 * @param {object} error : The exception sent back from the API.
 *
 * @returns {object}
 */
const deleteSubjectFailure = (error) => ({
	type: ActionTypes.DELETE_SUBJECT_FAILURE,
	payload: { error },
});

// //////////////////////////////////////////////////////// //
// //////////// Subject list clearing actions ///////////// //
// //////////////////////////////////////////////////////// //

/**
 * @function
 * @name clearSubjectListAction
 * @description Action triggered anytime the subject list is not needed anymore.
 *
 * @author Timothée Simon-Franza
 *
 * @returns {object}
 */
const clearSubjectListAction = () => ({ type: ActionTypes.CLEAR_SUBJECT_LIST });

// //////////////////////////////////////////////////////// //
// //////////////// Exported action creators ////////////// //
// //////////////////////////////////////////////////////// //

/**
 * @function
 * @name fetchSubject
 * @description Method used to fetch the latest version of a specific subject
 *
 * @author Timothée Simon-Franza
 *
 * @param {string} subjectId : The id of the subject we want to retrieve.
 */
export const fetchSubject = (subjectId) => (dispatch) => {
	dispatch(fetchSubjectRequest());

	return SubjectsApi.fetchSubjectById(subjectId)
		.then((subject) => dispatch(fetchSubjectSuccess({ subject })))
		.catch((error) => dispatch(fetchSubjectFailure(error)));
};

/**
 * @function
 * @name fetchSubjectList
 * @description Method used to fetch all subjects available to the current user.
 *
 * @author Timothée Simon-Franza
 */
export const fetchSubjectList = () => (dispatch) => {
	dispatch(fetchSubjectListRequest());

	return SubjectsApi.fetchSubjects()
		.then((subjects) => dispatch(fetchSubjectListSuccess({
			subjects,
			totalCount: subjects.length,
		})))
		.catch((error) => dispatch(fetchSubjectListFailure(error)));
};

/**
 * @function
 * @name createSubject
 * @description Method used to create a new subject in the database
 *
 * @author Timothée Simon-Franza
 *
 * @param {object} subjectData : The data to create the new subject from.
 */
export const createSubject = (subjectData) => (dispatch) => {
	dispatch(createSubjectRequest());

	return SubjectsApi.createSubject(subjectData)
		.then(({ subject }) => dispatch(createSubjectSuccess({ subject })))
		.catch((error) => dispatch(createSubjectFailure(error)));
};

/**
 * @function
 * @name updateSubject
 * @description Method used to update an existing subject instance from the database.
 *
 * @author Timothée Simon-Franza
 *
 * @param {object} subjectData	: The data to update the subject with.
 * @param {string} subjectId	: The id to identify the subject to update.
 */
export const updateSubject = (subjectData, subjectId) => (dispatch) => {
	dispatch(updateSubjectRequest());

	return SubjectsApi.updateSubject(subjectData, subjectId)
		.then(({ subject }) => dispatch(updateSubjectSuccess({ subject })))
		.catch((error) => dispatch(updateSubjectFailure(error)));
};

/**
 * @function
 * @name deleteSubject
 * @description Method used to remove an existing subject from the database.
 *
 * @author Timothée Simon-Franza
 *
 * @param {string} subjectId : The id identifying the subject to remove.
 */
export const deleteSubject = (subjectId) => (dispatch) => {
	dispatch(deleteSubjectRequest());

	return SubjectsApi.deleteSubject(subjectId)
		.then(({ subject }) => dispatch(deleteSubjectSuccess({ subject })))
		.catch((error) => dispatch(deleteSubjectFailure(error)));
};

/**
 * @function
 * @name clearSubjectList
 * @description Method used to clear the subject list stored in memory when the application does not need it anymore.
 *
 * @author Timothée Simon-Franza
 *
 * @returns {Promise.resolve}
 */
export const clearSubjectList = () => (dispatch) => {
	dispatch(clearSubjectListAction());

	return Promise.resolve();
};
