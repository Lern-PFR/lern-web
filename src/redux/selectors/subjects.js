import { sortBy } from 'lodash';
import { createSelector } from 'reselect';

/**
 * @function
 * @name getSubjects
 * @description A selector callback which returns all subjects from the Redux state.
 *
 * @author Timothée Simon-Franza
 *
 * @param {object} state The redux state at the time of call.
 *
 * @returns {Array}
 */
const getSubjects = (state) => state?.subjects?.items ?? { all: [], mine: [], active: [], available: [] };

/**
 * @function
 * @name getSubjectFilterValue
 * @description An intermidate selector callback used by the getSubjectsByTitleOrAuthor method to retrieve the filterValue param.
 *
 * @author Timothée Simon-Franza
 *
 * @param {string} filterValue The value to pass to the getSubjectsByTitleOrAuthor selector.
 *
 * @returns {string}
 */
const getSubjectFilterValue = (_, filterValue) => filterValue ?? '';

/**
 * @function
 * @name filterByTitleOrAuthor
 * @description Pure private method which filters the subjectList array param with the filterValue parameter's value.
 *
 * @author Timothée Simon-Franza
 *
 * @param {array} subjectList	The list of subjects to filter.
 * @param {string} filterValue	The value to use as a filtering criteria.
 *
 * @returns {Array}
 */
const filterByTitleOrAuthor = (subjectList, filterValue) => (
	subjectList.filter(({ title, author = undefined }) => (
		title.toLowerCase().includes(filterValue.toLowerCase())
		|| (author?.firstname && author.firstname.toLowerCase().includes(filterValue.toLowerCase()))
		|| (author?.lastname && author.lastname.toLowerCase().includes(filterValue.toLowerCase()))
	)) ?? []
);

/**
 * @function
 * @name getSubjectsByTitleOrAuthor
 * @description A selector callback which returns subjects whose name or author matches the filterValue param.
 *
 * @author Timothée Simon-Franza
 *
 * @param {string} filterValue The value to use for filtering.
 *
 * @returns {Array}
 */
const getSubjectsByTitleOrAuthor = createSelector(
	getSubjects,
	getSubjectFilterValue,
	({ mine, active, available }, filterValue) => ({
		mine: filterByTitleOrAuthor(mine, filterValue),
		active: filterByTitleOrAuthor(active, filterValue),
		available: filterByTitleOrAuthor(available, filterValue),
	})
);

/**
 * @function
 * @name getSubjectById
 * @description A selector callback which returns subjects whose name or author matches the subjectId param.
 *
 * @author Timothée Simon-Franza
 *
 * @param {string} subjectId The id of the subject we want to retrieve.
 *
 * @returns {object|undefined}
 */
const getSubjectById = createSelector(
	getSubjects,
	(_, subjectId) => subjectId,
	({ all: subjects }, subjectId) => {
		const subject = subjects.filter(({ id }) => (id === subjectId))?.[0] ?? undefined;

		if (subject === undefined) {
			return undefined;
		}

		return {
			...subject,
			modules: sortBy(subject?.modules || [], 'order'),
		};
	}
);

export {
	getSubjects,
	getSubjectById,
	getSubjectsByTitleOrAuthor,
};
