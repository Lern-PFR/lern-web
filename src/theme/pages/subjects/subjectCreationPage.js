import { coconut, lightgrey, oyster, peppercorn, tuna } from 'theme/colors';
import { mainLayoutTopPadding } from 'theme/constants';
import { doublePica } from 'theme/textStyles';

const layout = {
	display: 'grid',
	gridTemplateColumns: '25% 1fr',
	gridGap: '2em',
	paddingRight: mainLayoutTopPadding,
	height: '100%',
};

const sidebar = {
	backgroundColor: lightgrey,
	boxShadow: `0 2px 4px ${peppercorn.default}29`,
	margin: 0,
	padding: 0,
	paddingTop: mainLayoutTopPadding,
	listStyle: 'none',
};

const sidebarElement = {
	...doublePica,
	color: tuna.default,
	padding: '.3em 1em',
	cursor: 'pointer',
	lineHeight: 'auto',
	transition: 'color .1s ease-in-out',

	'&:hover': {
		color: peppercorn.default,
	},
};

const sidebarCurrentElement = {
	color: coconut.default,
	backgroundColor: oyster.default,
	padding: 0,

	'&:hover': {
		color: coconut.default,
	},
};

const contentSection = {
	paddingTop: mainLayoutTopPadding,
};

const title = {
	lineHeight: 'auto',
	margin: 0,
	paddingBottom: '24px',
};

const form = {
	display: 'grid',
	gridGap: '1em',
	height: 'max-content',
};

const formSubmit = {
	justifySelf: 'start',
};

export {
	layout,
	sidebar,
	sidebarElement,
	sidebarCurrentElement,
	contentSection,
	title,
	form,
	formSubmit,
};
