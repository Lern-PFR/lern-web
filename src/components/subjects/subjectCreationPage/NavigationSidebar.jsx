import { StyledList } from 'components/shared/styledElements';
import { useTranslation } from 'react-i18next';
import { sidebar, sidebarCurrentElement, sidebarElement } from 'theme/pages/subjects/subjectCreationPage';
import styled from 'styled-components';

const SidebarElement = styled('li')(
	{
		...sidebarElement,
		color: (({ isCurrent }) => (isCurrent ? sidebarCurrentElement.color : sidebarElement.color)),
		backgroundColor: (({ isCurrent }) => (isCurrent ? sidebarCurrentElement.backgroundColor : sidebarElement.backgroundColor)),

		'&:hover': {
			color: ({ isCurrent }) => (isCurrent ? sidebarCurrentElement['&:hover'].color : sidebarElement['&:hover'].color),
		},
	},
);

/**
 * @name NavigationSidebar
 * @description A sidebar component used in the content creation and edition pages to navigate between a subject's children.
 *
 * @author TimothÃ©e Simon-Franza
 */
const NavigationSidebar = () => {
	const { t } = useTranslation();

	return (
		<StyledList {...sidebar}>
			<SidebarElement isCurrent>{t('subjects.creation.sidebar.new_subject')}</SidebarElement>
		</StyledList>
	);
};

export default NavigationSidebar;
