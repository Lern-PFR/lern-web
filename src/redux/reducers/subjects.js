import { ActionTypes } from 'redux/actions/subjects';

const initialState = {
	isLoading: false,
	items: {
		// Used most of the time
		all: [], 		// Contains all subjects sent back from the API.

		// Used when we need to separate by category
		mine: [],		// Contains subjects created by the current user.
		active: [],		// Contains subjects currently active for the current user.
		available: [],	// Contains the rest of valid subjects.
	},
	totalCount: 0,
};

export default (state = initialState, { type, payload }) => {
	switch (type) {
		case ActionTypes.FETCH_SUBJECT_REQUEST:
		case ActionTypes.FETCH_SUBJECT_LIST_REQUEST:
		case ActionTypes.CREATE_SUBJECT_REQUEST:
		case ActionTypes.UPDATE_SUBJECT_REQUEST:
		case ActionTypes.DELETE_SUBJECT_REQUEST:
			return {
				...state,
				isLoading: true,
			};
		case ActionTypes.FETCH_SUBJECT_FAILURE:
		case ActionTypes.FETCH_SUBJECT_LIST_FAILURE:
		case ActionTypes.CREATE_SUBJECT_FAILURE:
		case ActionTypes.UPDATE_SUBJECT_FAILURE:
		case ActionTypes.DELETE_SUBJECT_FAILURE:
			return {
				...state,
				isLoading: false,
			};
		case ActionTypes.CREATE_SUBJECT_SUCCESS:
		case ActionTypes.UPDATE_SUBJECT_SUCCESS:
		case ActionTypes.DELETE_SUBJECT_SUCCESS:
			return {
				...state,
				isLoading: false,
			};
		case ActionTypes.FETCH_SUBJECT_SUCCESS:
			return {
				...state,
				isLoading: false,
				items: {
					...initialState.items,
					all: [payload.subject],
				},
				totalCount: 1,
			};
		case ActionTypes.FETCH_SUBJECT_LIST_SUCCESS:
			return {
				...state,
				isLoading: false,
				items: {
					all: [],
					...payload.subjects,
				},
				totalCount: payload.totalCount,
			};
		case ActionTypes.CLEAR_SUBJECT_LIST:
			return {
				...state,
				items: initialState.items,
				totalCount: 0,
			};
		default:
			return state;
	}
};
