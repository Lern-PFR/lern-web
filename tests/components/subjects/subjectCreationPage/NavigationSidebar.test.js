import { shallow } from 'enzyme';
import { NavigationSidebar } from 'components/subjects/subjectCreationPage';

describe('NavigationSidebar', () => {
	it('should match previous snapshot', () => {
		const sut = (<NavigationSidebar />);
		const wrapper = shallow(sut);

		expect(wrapper).toMatchSnapshot();
	});
});
