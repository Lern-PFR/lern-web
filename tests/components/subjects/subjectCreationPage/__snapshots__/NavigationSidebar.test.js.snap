// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavigationSidebar should match previous snapshot 1`] = `
<styled.ul
  backgroundColor="#FBFCFD"
  boxShadow="0 2px 4px #241C1529"
  listStyle="none"
  margin={0}
  padding={0}
  paddingTop={64}
>
  <styled.li
    isCurrent={true}
  >
    subjects.creation.sidebar.new_subject
  </styled.li>
</styled.ul>
`;
